core/finality-grandpa/src/tests.rs

```diff
@@ -569,28 +466,30 @@ fn finalize_3_voters_no_observers() {
 
 #[test]
 fn finalize_3_voters_1_full_observer() {
+	let mut runtime = current_thread::Runtime::new().unwrap();
+
 	let peers = &[AuthorityKeyring::Alice, AuthorityKeyring::Bob, AuthorityKeyring::Charlie];
 	let voters = make_ids(peers);
 
 	let mut net = GrandpaTestNet::new(TestApi::new(voters), 4);
 	net.peer(0).push_blocks(20, false);
-	net.sync();
+	runtime.block_on(futures::future::poll_fn::<(), (), _>(|| Ok(net.poll_until_sync()))).unwrap();
```

Comment:
Would rather see . `poll_until_sync' accept runtime parameter and calling `block_on` internally, to avoid repeating all the `poll_fn` ugliness. A more generic `poll_until` that accepts a function would also be useful.